### ISSUE TITLE
Integrate with dependent: :delete_all

### ISSUE THREAD

**Issue by nornagon** [rubysherpas/paranoia#68]
Currently specifying a relation with `dependent: :delete_all` hard-deletes the dependent rows instead of just setting their `deleted_at`.

**Comment by radar** [comment]
This is expected behaviour due to how Active Record works. It will issue a DELETE query for the association rather than calling `destroy` on each object. I do not know how to fix this. Could you please investigate it and submit a patch?

**Comment by korbin** [comment]
ActsAsParanoid implemented this by overloading the delete_all method on the Arel relation; should this exist in Paranoia as well?

**Comment by radar** [comment]
I'd rather we didn't monkey-patch things within paranoia, especially ARel code. It's supposed to be a light library for deletion. I don't want to maintain this huge complex system [as I already help maintain not one](https://github.com/spree/spree) [but two](http://github.com/activerecord-hackery/ransack).

**Comment by korbin** [comment]
Agreed - I like the opt-in functionality.

I looked into adding another :soft_delete_all method, but it still involved monkey-patching

https://github.com/rails/rails/blob/master/activerecord/lib/active_record/associations/builder/has_many.rb

to add another method.

---

For posterity and solution (I can submit a readme PR if you like, untested):

``` ruby
before_destroy :soft_delete_dependent_widgets

def soft_delete_dependent_widgets
  self.widgets.update_all(deleted_at: current_time_from_proper_timezone)
end
```

**Comment by radar** [comment]
@korbin Yes please do submit a PR for this.

**Comment by rorhug** [comment]
Has anything happened with this yet?

It seems really inefficient if a parent has, say, 100 child records and to `UPDATE` each one separately on deletion of the parent. That's why Rails has `delete_all`. It would be very useful to see it here too.

**Comment by radar** [comment]
@RoryDH Nope, nothing has happened. If you think this would be useful, please submit a patch.

**Comment by salmanasiddiqui** [comment]
+1 on this.
I am currently adding soft delete thing in my app, reviewed few gems and went with this one.
But im stuck on this issue aswell, there should be a method to soft delete in bulk. Infact there should be a restore_all method aswell.. and also a way to skip restore callbacks.

for example, `restore` & `restore_all` should execute the restore callbacks, just like `destroy` & `destroy_all`
and `restore!` & `restore_all!` shouldn't execute restore callbacks, similar to `delete` & `delete_all`

or may be use `recover` & `recover_all` instead of putting the `!` mark.. To make it more friendly to Rails community.

Im kinda short on time otherwise would have submitted a PR.

Guess for now I have to use `update_all(deleted_at: Time.now)` instead of `delete_all` and `update_all(deleted_at: nil)` to mimic `restore_all`

**Comment by fzondlo** [comment]
Here's a monkey patch solution to override the AR delete_all method:

```
module ActiveRecord
  class Relation
    def delete_all
      self.update_all(deleted_at: Time.now)
    end
  end
end
```

**Comment by rwojsznis** [comment]
Hi, I'm aware this ticket is old as hell, but I peeked into active record source and indeed hooking into `delete_all` would require a lot of monkey patching as rails is simply building&executing arel query under the hood :confused: Maybe (big maybe) introducing some custom `dependent: :paranoia_delete_all` method  would be easier (again so monkey patching would be required by looking at code, but should be less invasive) - so at least soft deleting associations would work nicely.

@radar - Wdyt about such approach?
